fury/stream/tools.py

```
@@ -0,0 +1,1000 @@
+from enum import auto
+import threading
+import numpy as np
+import io
+from PIL import Image, ImageDraw
+import multiprocessing
+import time
+import logging
+from threading import Timer
+import asyncio
+from abc import ABC, abstractmethod
+from fury.stream.constants import PY_VERSION_8
+
+if PY_VERSION_8:
+    from multiprocessing import shared_memory
+    from multiprocessing import resource_tracker
+else:
+    shared_memory = None
+
+try:
+    import cv2
+    OPENCV_AVAILABLE = True
+except ImportError:
+    cv2 = None
+    OPENCV_AVAILABLE = False
+
+_FLOAT_ShM_TYPE = 'd'
+_INT_ShM_TYPE = 'i'
+_UINT_ShM_TYPE = 'I'
+_BYTE_ShM_TYPE = 'B'
+
+_FLOAT_SIZE = np.dtype(_FLOAT_ShM_TYPE).itemsize
+_INT_SIZE = np.dtype(_INT_ShM_TYPE).itemsize
+_UINT_SIZE = np.dtype(_UINT_ShM_TYPE).itemsize
+_BYTE_SIZE = np.dtype(_BYTE_ShM_TYPE).itemsize
+
+
+def remove_shm_from_resource_tracker():
+    """Monkey-patch multiprocessing.resource_tracker so SharedMemory won't
+    be tracked
+
+    Notes
+    -----
+        More details at: https://bugs.python.org/issue38119
+
+    """
+
+    def fix_register(name, rtype):
+        if rtype == "shared_memory":
+            return
+        return resource_tracker._resource_tracker.register(
```

`self` is undefined here. Results in a silent error when the server is running in another process.

---

This is how we can do monkey patching on python. This is an open bug in CPython implementation regarding shared memory

---

Alto this is silent ERROR or silent Warning?

---

Indeed this results in a warning caused by an unraisable exception, which I'm not sure if it is unraisable because it is running during garbage collector, destroy or because it runs in another process.

---

What you can do is call `ls -ls /dev/shm` and see if some resource was not properly removed from.

---

Ok, I've tested this, and the patch works on mac. The problem is that, for some reason, pytest checks this self and gives an NameError. On macs, /dev/shm does not exist, but I've checked for memory usage and everything seems to work nicely on Python 3.8.

---

Adding a NameError try catch, as you suggested, gets rid of the warning for when running pytest. Memory is reclaimed. so everything should be fine. The problem does not happen on Windows.

---

Done at https://github.com/fury-gl/fury/pull/437/commits/25b39792d69c212eb02c8bd5f6957c50cee3c95e
Maybe is good to describe this issue on cypthon repo. https://github.com/python/cpython/issues/82300